rpc: touch up DialBack functionality

tbg opened the pull request:
Extracted from https://github.com/cockroachdb/cockroach/pull/99191, where I had to touch the test to actually pass with the stateful breakers. (The actual test refactor is in #99191).

----


Clarified the semantics a bit and made sure it didn't dig into the lower layers
more than it needed to. Left some comments around oddities of the approach that
are good to know when working in the area.

Epic: none
Release note: None

tbg committed:
rpc: touch up DialBack functionality

Clarified the semantics a bit and made sure it didn't dig into the lower layers
more than it needed to. Left some comments around oddities of the approach that
are good to know when working in the area.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103683)
<!-- Reviewable:end -->

tbg committed:
rpc: inline and rename dialback attempt

Make it clear that this is only related to dialbacks,
and remove an unnecessary callback.

Epic: none
Release note: None

tbg committed:
rpc: expand comment on VerifyDialback

Took me a bit to grok it, shouldn't be so hard for
the next person.

Also, I think I can simplify this after #99191.

Epic: none
Release note: None

tbg committed:
rpc: fix data race in TestHeartbeatDialback

It was *writing into* ErrNoConnection rather than just checking the type.

Epic: none
Release note: None

tbg commented:
cc @aliher1911 since you'll be reviewing https://github.com/cockroachdb/cockroach/pull/99191, and @erikgrinaker since you reviewed the original DialBack PRs and fix-ups such as https://github.com/cockroachdb/cockroach/issues/99104.

aliher1911 commented on pkg/rpc/context.go:
```diff
@@ -2665,14 +2662,47 @@ func (rpcCtx *Context) VerifyDialback(
+		// (because we only check "our" map, not the rpcCtx's map).
+		//
+		// None of this causes problems today because the remote always starts its
+		// heartbeat loop with a BLOCKING ping, follower by NON_BLOCKING pings until
```
s/follower/followed/

aliher1911 reviewed (approved):
I think the lack of overview for dialback is making it hard to easily understand. E.g. how it works in good case and when connection is dropped. e.g. Node A dials Node B first time, Node B dials Node A when there's already a connection, Connection from A to B fails while B talks to A etc. And a lifecycle of connection with respect of pings. Then cases like why we keep sticky error in dialback map would make more sense. 
I tried to search if there are comments referring to dialback types other than in type itself, and didn't see any. But maybe I just missed something.

tbg commented:
Good point, I'll backfill some more of the documentation, will ping you when it's ready.

andrewbaptist reviewed (changes requested):
We should chat about this briefly. The changes you made all see useful, but the comments don't seem accurate. Either I goofed something up in this implementation, or it works different than the comments. I can try and write a short "overview" of how this is intended to work as Pavel suggested if you haven't already started.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-:-NW8A-dpEodhvTXxvXSH:b9bmfdq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[`pkg/rpc/context.go` line 2678 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW88kDw5fGMezvTmrkn:-NW88kDw5fGMezvTmrko:bwg2itc) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2678)):*
> ```Go
> 		// and we remove the map entry.
> 		//
> 		// The next call will then insert the connection again, following the same
> ```

Add "If the connection is still unhealthy". We don't insert if the connection is currently healthy because of the previous health check before we get here.
___
*[`pkg/rpc/context.go` line 2681 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW891xb9WyJ-hwL7bqg:-NW891xb9WyJ-hwL7bqh:b55lore) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2681)):*
> ```Go
> 		// logic. (It is likely still healthy).
> 		//
> 		// The next call will then remove it again (since it was healthy), and so on.
> ```

This doesn't seem right, we don't insert/remove on healthy connections, only on connections where we are unable to dialback.
___
*[`pkg/rpc/context.go` line 2685 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW88YdnAjl0rTrerhdG:-NW88YdnAjl0rTrerhdH:by2ti6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2685)):*
> ```Go
> 		// At some point, we might find that the connection is unhealthy. It is then
> 		// kept in the map and all NON_BLOCKING pings bounce off this stored
> 		// connection. It will not be reset until a BLOCKING ping arrives from the
> ```

Or this node manually creates a connection back to the originating node.
___
*[`pkg/rpc/context.go` line 2697 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW89EBICgK5yKUkdMRR:-NW89EBICgK5yKUkdMRS:brui6vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2697)):*
> ```Go
> 		// TODO(tbg): this structure is odd: BLOCKING and NON_BLOCKING are tightly
> 		// coupled which is surprising and could lead to bugs if we made changes to
> 		// the callers. Also, the alternating addition and removal of a healthy
> ```

This doesn't happen because we don't get here in the code once the connection is healthy, see notes above.
___
*[`pkg/rpc/context.go` line 2750 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW89Qqe42uPSliFG9S2:-NW89Qqe42uPSliFG9S3:b-q5onlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2750)):*
> ```Go
> 	// protection we would continually try opening new dialback connections, but
> 	// never observe the result.
> 	err := previousAttempt.Health()
> ```

nit: This is a good change to use Health, maybe just make a note on the Health function that if its behavior changes, we should make sure this is still correct. The current behavior of Health is not completely obvious. 
___
*[`pkg/rpc/heartbeat.proto` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW89yut3BqPgtlVcLcK:-NW89yut3BqPgtlVcLcL:b7km3ku) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/heartbeat.proto#L83)):*
> ```protobuf
>     // failure to the next PingRequest on this connection. This failure is
>     // then permanent (even if a redial would succeed) until the caller
>     // uses a BLOCKING heartbeat to kick off another attempt.
> ```

This isn't right.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Thanks! You were right that the comments had some holes. I updated them, most of the points are still there but the create-remove loops don't actually occur, this was based on me misremembering what the code actually does when I wrote them (I think I thought the blocking attempt was also clearing out the map, which it does not).

> Dismissed @aliher1911 and @andrewbaptist from 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-:-NWBuqDi3ei1T-e1H2eu:bh2b93k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @andrewbaptist)

___
*[`pkg/rpc/context.go` line 2678 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW88kDw5fGMezvTmrkn:-NWBt5CP4M0XidU7aRrc:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2678)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Add "If the connection is still unhealthy". We don't insert if the connection is currently healthy because of the previous health check before we get here.
</blockquote></details>

Removed.
___
*[`pkg/rpc/context.go` line 2681 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW891xb9WyJ-hwL7bqg:-NWBtq9x6zRmHaHO2UfS:b-1h1rck) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2681)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This doesn't seem right, we don't insert/remove on healthy connections, only on connections where we are unable to dialback.
</blockquote></details>

You are right, removed.
___
*[`pkg/rpc/context.go` line 2685 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW88YdnAjl0rTrerhdG:-NWBu-9K1suIb6xw6RX4:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2685)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Or this node manually creates a connection back to the originating node.
</blockquote></details>

Removed.
___
*[`pkg/rpc/context.go` line 2692 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#gh-1202084521:-NWBu0Of2TYz--DdRvjS:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2692)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

s/follower/followed/
</blockquote></details>

Removed.
___
*[`pkg/rpc/context.go` line 2697 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW89EBICgK5yKUkdMRR:-NWBu1P8D4Hui5X6f3QF:b-tgmopg) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2697)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This doesn't happen because we don't get here in the code once the connection is healthy, see notes above.
</blockquote></details>

I updated this, PTAL.
___
*[`pkg/rpc/context.go` line 2750 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW89Qqe42uPSliFG9S2:-NWBtSIk87cf375GRssU:b-2g33la) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/context.go#L2750)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

nit: This is a good change to use Health, maybe just make a note on the Health function that if its behavior changes, we should make sure this is still correct. The current behavior of Health is not completely obvious. 
</blockquote></details>

Which part of the behavior do you mean? That it returns `ErrNotHeartbeated` when the first connection attempt has not concluded yet? That part is documented already.
___
*[`pkg/rpc/heartbeat.proto` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103683#-NW89yut3BqPgtlVcLcK:-NWBu65g6kB0rLl_9Sa0:bmbdxnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/925cddfce891eeb382940bd7111120fc2d9c1407/pkg/rpc/heartbeat.proto#L83)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This isn't right.
</blockquote></details>

I believe this is right. We're storing a `*Connection` into the map, which is really a specific connection attempt. If the "fast-path" system class dialback connection just never gets attempted, once we have an errored entry in the map, nothing will ever remove it. The only thing that will save us at that point is a BLOCKING ping (which we do issue, so it's "fine", just a bit more subtle than it needs to be).

This whole machinery can be boiled down to ~nothing once we have https://github.com/cockroachdb/cockroach/pull/99191, so I think we should just do that (rather than building a parallel ad-hoc circuit breaker), but I left an alternative in the comments just in case.


<!-- Sent from Reviewable.io -->

tbg commented:
@andrewbaptist friendly ping, this is still in the critical path of #99191 which I hope will be green soon-ish. If you have substantial suggestions, please let me know which of those you're comfortable having me do as a follow-up. I'd like to keep the critical path minimal. Thanks!

andrewbaptist commented:
I added this doc (https://docs.google.com/document/d/1MOZSCIMhRHvvbYu-EWAlokqQaH1QRfJ0LuKK_1QUygs/edit) - can you incorporate this into the doc PR as an overview. Feel free to edit it. Let me know if you have any questions.

tbg commented:
Thanks Andrew! I'm noticing that we disagree about 3) in the doc. I think that if you use *only* NON_BLOCKING and then a dialback attempt fails (and the SysClass connection is never ready, perhaps because it failed once and then was never recreated), the result of that failed attempt will be sticky, i.e. the connection will be permanently defunct.

In the interest of getting this PR in, would you mind delegating the doc PR to a follow-up (which I'm happy to open right as this goes in)?

And see my previous question, could you green-light this PR or let me know if there is anything else to address first?

andrewbaptist commented:
:lgtm:

tbg commented:
bors r=andrewbaptist
